File: src/HARD/664. Strange Printer.ts
// 04/19/2021 HARD

// https://leetcode-cn.com/problems/strange-printer/

/*
There is a strange printer with the following two special requirements:

The printer can only print a sequence of the same character each time.
At each turn, the printer can print new characters starting from and ending at any places,
and will cover the original existing characters.
Given a string consists of lower English letters only, 
your job is to count the minimum number of turns the printer needed in order to print it.

Example 1:
Input: "aaabbb"
Output: 2
Explanation: Print "aaa" first and then print "bbb".
Example 2:
Input: "aba"
Output: 2
Explanation: Print "aaa" first and then print "b" from the second place of the string, which will cover the existing character 'a'.
Hint: Length of the given string will not exceed 100.

 */

function strangePrinter(s: string): number {
  if (!s.length) return 0;
  const dp: number[][] = new Array(s.length);
  for (let i = 0; i < s.length; i++) {
    dp[i] = new Array(s.length).fill(0);
    dp[i][i] = 1;
  }
  for (let i = s.length - 2; i >= 0; i--) {
    for (let j = i + 1; j < s.length; j++) {
      dp[i][j] = dp[i][j - 1] + 1;
      for (let k = i; k < j; k++) {
        if (s[k] === s[j]) {
          dp[i][j] = Math.min(dp[i][j], dp[i][k] + dp[k + 1][j - 1]);
        }
      }
    }
  }
  return dp[0][s.length - 1];
};

File: src/HARD/664-StrangePrinter.cpp
//
// Created by ZiqianCheng on 2021/5/24.
//

// HARD https://leetcode-cn.com/problems/strange-printer/

/*
 * There is a strange printer with the following two special properties:

The printer can only print a sequence of the same character each time.
At each turn, the printer can print new characters starting from and ending at any place and will cover the original existing characters.
Given a string s, return the minimum number of turns the printer needed to print it.

 

Example 1:

Input: s = "aaabbb"
Output: 2
Explanation: Print "aaa" first and then print "bbb".
Example 2:

Input: s = "aba"
Output: 2
Explanation: Print "aaa" first and then print "b" from the second place of the string, which will cover the existing character 'a'.
 

Constraints:

1 <= s.length <= 100
s consists of lowercase English letters.
 */

#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
class Solution {
public:
  int strangePrinter(string s) {
    int n = s.size();
    vector<vector<int>> dp(n, vector<int>(n, 0));
    for (int i = 0; i < n; ++i) dp[i][i] = 1;
    for (int i = n - 1; i >= 0; --i) {
      for (int j = i + 1; j < n; ++j) {
        if (s[i] == s[j]) dp[i][j] = dp[i][j - 1];
        else {
          int minValue = INT_MAX;
          for (int k = i; k < j; ++k) {
            minValue = min(minValue, dp[i][k] + dp[k + 1][j]);
          }
          dp[i][j] = minValue;
        }
      }
    }
    return dp[0][n - 1];
  }
};

int main() {
  Solution s;
  cout << s.strangePrinter("aaabbb") << endl;
  cout << s.strangePrinter("aba") << endl;
  return 0;
}